crates/turborepo-auth/src/lib.rs

```diff
@@ -12,19 +12,139 @@ mod ui;
+    pub fn into_inner(&self) -> &str {
         match self {
-            Self::Existing(s) => s,
-            Self::New(s) => s,
+            Self::Existing(token) | Self::New(token) => token.as_str(),
+        }
+    }
+}
+
+fn current_unix_time() -> u128 {
+    use std::time::{SystemTime, UNIX_EPOCH};
+    SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .expect("Time went backwards")
+        .as_millis()
+}
+
+// As of the time of writing, this should always be true, since a token that
+// isn't active returns an error when fetching metadata for the token.
+fn is_token_active(metadata: &ResponseTokenMetadata) -> bool {
+    let current_time = current_unix_time();
+    let active_at = metadata.active_at;
+
+    let active_less_than_current = active_at <= current_time;
+    let all_scopes_active = metadata.scopes.iter().all(|scope| {
+        scope
+            .expires_at
+            .map_or(true, |expires_at| expires_at > current_time)
+    });
+
+    // Not all scopes have an expiration date, so we need to check if all of them
+    // are expired. If there isn't an expiration date, we assume they are infinite
+    // and therefore cannot be expired.
+    active_less_than_current && all_scopes_active
+}
+
+#[cfg(test)]
+mod tests {
+    use turborepo_vercel_api::token::Scope;
+
+    use super::*;
+
+    #[test]
+    fn test_is_token_active() {
+        let current_time = current_unix_time();
+        let quick_scope = |expiry| Scope {
+            expires_at: expiry,
+            scope_type: "".to_string(),
+            origin: "".to_string(),
+            created_at: 0,
+            team_id: None,
+        };
+        let mock_response = |active_at, scopes| ResponseTokenMetadata {
+            active_at,
+            scopes,
+            // These fields don't matter in the test
+            id: "".to_string(),
+            name: "".to_string(),
+            token_type: "".to_string(),
+            origin: "".to_string(),
+            created_at: 0,
+        };
+
+        let cases = vec![
+            // Case: Token active, no scopes (implicitly infinite)
+            (current_time - 10, vec![], true),
+            // Case: Token active, one scope without expiration
+            (current_time - 10, vec![quick_scope(None)], true),
+            // Case: Token active, one scope expired
+            (
+                current_time - 10,
+                vec![quick_scope(Some(current_time - 1))],
+                false,
+            ),
+            // Case: Token active, one scope not expired
+            (
+                current_time - 10,
+                vec![quick_scope(Some(current_time + 11))],
+                true,
+            ),
+            // Case: Token inactive (future `active_at`)
+            (
+                current_time + 10,
+                vec![quick_scope(Some(current_time + 20))],
+                false,
+            ),
+        ];
+
+        for (active_at, scopes, expected) in cases {
```

Comment:
I do think using [test_case](https://docs.rs/test-case/latest/test_case/) instead of manual table tests has some benefits, primarily it'll run all of the test cases on failure and not bail after the first failure. 

Not blocking, just sharing my 2 cents

Reply:
yeah I've been experimenting with different ways to write tests, and originally this was a `tokio::test` which I wasn't sure of how it'd interact with `test_case`, so I went back to my old Go route. I'll probably make this use `test_case` in a future PR